sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -156,14 +159,16 @@ gather(const Tx *p, simd<uint32_t, N> offsets, simd_mask<N> mask = 1) {
 /// @tparam N Number of elements to write; can be \c 1, \c 2, \c 4, \c 8, \c 16
 ///   or \c 32.
 /// @param p The base address.
-/// @param offsets A vector of 32-bit offsets in bytes. For each lane \c i,
-///   ((byte*)p + offsets[i]) must be element size aligned.
+/// @param offsets A vector of 32-bit or 64-bit offsets in bytes. For each lane
+/// \c i,   ((byte*)p + offsets[i]) must be element size aligned.
 /// @param vals The vector to scatter.
 /// @param mask The access mask, defaults to all 1s.
 ///
-template <typename Tx, int N, class T = detail::__raw_t<Tx>>
-__ESIMD_API std::enable_if_t<detail::isPowerOf2(N, 32)>
-scatter(Tx *p, simd<uint32_t, N> offsets, simd<Tx, N> vals,
+template <typename Tx, int N, class T = detail::__raw_t<Tx>, typename Toffset>
+__ESIMD_API std::enable_if_t<detail::isPowerOf2(N, 32) &&
+                             (std::is_same_v<Toffset, uint32_t> ||
```

Comment:
I believe this can break existing code. E.g. int offsets will now cause compilation error, but should have been working in the past.

Reply:
Backward compatibility of non-experimental code must not be broken. This is a hard requirement. The only exception is approved list of APIs changed in major releases.

> The only code breaking problem I saw so far is when offset is a simd_view i.e. something like:
scatter(p, x.select<1,8>(0), vals)

that is a problem too.
I tried `short` as offset type for scatter - it works now. So it should continue to work. My suggestion is:

1) create `template<class OffsetT> ... scatter_impl(..., simd<OffsetT, N> offsets,...)` in namespace detail.
2) create `...scatter(Tx *p, simd<uint64_t, N> offsets,...`
3) have `...scatter(Tx *p, simd<uint64_t, N> offsets,...` and `...scatter(Tx *p, simd<uint32_t, N> offsets,...` delegate to scatter_impl.

> I am not sure that the code we have now is good one as it would accept even vector of floats without any issue (I have 2 tests where offsets and vals were flipped and did not cause any compilation issue

Implicit conversion between arithmetic types in most contexts is basic C++ behavior, `esimd::simd` tries to follow. The user logic error you mention is desirable to protect against, of course, but even the enable_if_t you added does not fully shield from this if element type is uint32/64.